[Android] Allow to use system fonts

jsuarezruiz opened the pull request:
### Description of Change

Allow to use Android system fonts directly.

```
<VerticalStackLayout VerticalOptions="Center">
    <Label
        Text="Hello world" />
    <Label
        FontFamily="monospace"
        Text="Hello world" />
    <Label
        FontFamily="serif"
        Text="Hello world" />
</VerticalStackLayout>
```

![Screenshot_1683643513](https://github.com/dotnet/maui/assets/6755973/c36b1c1f-fd96-4be2-83e6-5c6a5913f788)
![image](https://github.com/dotnet/maui/assets/6755973/f716a725-b780-41fd-a01e-6cc6c4014f66)

### Issues Fixed

Fixes #11422

jsuarezruiz committed:
Allow to use Android system fonts directly

jsuarezruiz committed:
Added some tests

mattleibow commented on src/Core/src/Fonts/FontManager.Android.cs:
```diff
@@ -130,6 +135,21 @@ public FontSize GetFontSize(Font font, float defaultFontSize = 0)
+			{
+				case "monospace":
+					return Typeface.Monospace;
+				case "sansserif":
```
Should we add some "sans-serif" - and any CSS equivalents? What would you do in Native Android / CSS to get these fonts? I think `sans-serif` at least?

mattleibow commented on src/Core/src/Fonts/FontManager.Android.cs:
```diff
@@ -130,6 +135,21 @@ public FontSize GetFontSize(Font font, float defaultFontSize = 0)
+			}
+		}
+
 		Typeface? CreateTypeface((string? fontFamilyName, FontWeight weight, bool italic) fontData)
```
I wonder if this method is the correct place? `CreateTypeface`

The `LoadTypefaceFromAsset` is very specific so maybe moving this to check before - in the caller?

jsuarezruiz replied:
Good point, I have reviewed the code and I think the changes are now in the right place.

jsuarezruiz committed:
Merge branch 'main' into fix-11422

jsuarezruiz committed:
Updated impl

Commit:
Auto-format source code

jonathanpeppers commented:
@jsuarezruiz these changes cause an exception to be thrown several times on startup in an app I've been testing:

```
06-19 15:31:19.834  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium' from assets.
06-19 15:31:19.840  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium.ttf' from assets.
06-19 15:31:19.840  5706  5706 I DOTNET  : Unable to load font 'Fonts/sans-serif-medium.ttf' from assets.
06-19 15:31:19.841  5706  5706 I DOTNET  : Unable to load font 'fonts/sans-serif-medium.ttf' from assets.
06-19 15:31:19.841  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium.otf' from assets.
06-19 15:31:19.842  5706  5706 I DOTNET  : Unable to load font 'Fonts/sans-serif-medium.otf' from assets.
06-19 15:31:19.842  5706  5706 I DOTNET  : Unable to load font 'fonts/sans-serif-medium.otf' from assets.
```

https://github.com/dotnet/maui/blob/503b5344c40e49b234c0c7a654881250ecdb550e/src/Core/src/Fonts/FontManager.Android.cs#L127-L128

On a Pixel 5, `dotnet-trace` shows a decent amount of time doing this:

```
64.98ms  microsoft.maui!Microsoft.Maui.Handlers.LabelHandler.MapFont(Microsoft.Maui.Handlers.ILabelHandler,Microsoft.Maui.ILabel)
```

I open the Shell flyout in this app to see these:

https://github.com/supershopping/ShellFlyoutLagging

Is there something going wrong here?

jsuarezruiz commented:
> @jsuarezruiz these changes cause an exception to be thrown several times on startup in an app I've been testing:
> 
> ```
> 06-19 15:31:19.834  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium' from assets.
> 06-19 15:31:19.840  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium.ttf' from assets.
> 06-19 15:31:19.840  5706  5706 I DOTNET  : Unable to load font 'Fonts/sans-serif-medium.ttf' from assets.
> 06-19 15:31:19.841  5706  5706 I DOTNET  : Unable to load font 'fonts/sans-serif-medium.ttf' from assets.
> 06-19 15:31:19.841  5706  5706 I DOTNET  : Unable to load font 'sans-serif-medium.otf' from assets.
> 06-19 15:31:19.842  5706  5706 I DOTNET  : Unable to load font 'Fonts/sans-serif-medium.otf' from assets.
> 06-19 15:31:19.842  5706  5706 I DOTNET  : Unable to load font 'fonts/sans-serif-medium.otf' from assets.
> ```
> 
> https://github.com/dotnet/maui/blob/503b5344c40e49b234c0c7a654881250ecdb550e/src/Core/src/Fonts/FontManager.Android.cs#L127-L128
> 
> (NOTE: I changed the above to `Console.WriteLine()` and fixed the missing `$` locally)
> 
> On a Pixel 5, `dotnet-trace` shows a decent amount of time doing this:
> 
> ```
> 64.98ms microsoft.maui!Microsoft.Maui.Handlers.LabelHandler.MapFont(Microsoft.Maui.Handlers.ILabelHandler,Microsoft.Maui.ILabel)
> 38.92ms microsoft.maui!Microsoft.Maui.FontManager.LoadTypefaceFromAsset(string,bool)
> ```
> 
> I open the Shell flyout in this app to see these:
> 
> https://github.com/supershopping/ShellFlyoutLagging
> 
> Is there something going wrong here?

I think you found this one: https://github.com/dotnet/maui/issues/13239
Searching the possible origin, could be:

https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/Platform/Android/BottomNavigationViewUtils.cs#L218
https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellItemRenderer.cs#L248

jonathanpeppers commented:
Ok, I think we should add a couple more cases: https://github.com/dotnet/maui/pull/15759

It seems to help performance to avoid these lookups.

ewerspej commented:
Is there an equivalent for iOS? I'm trying to use an iOS-specific font in a project that has been converted from XF to MAUI, like this:

```xml
<Style x:Key="bodySmallSemibold" TargetType="Label">
    <Setter Property="FontFamily">
        <Setter.Value>
            <OnPlatform x:TypeArguments="x:String">
                <On Platform="iOS">.SFUI-Semibold</On>
                <On Platform="Android">sans-serif-medium</On>
            </OnPlatform>
        </Setter.Value>
    </Setter>
</Style>
```

But the font is not used. Should I open a ticket for this or is this already being worked on/known?
